File: bst.cpp
/*
 * binary search tree
 * github.com/01mu
 */

#include "bst.h"

#ifndef BST_CPP_
#define BST_CPP_

template <class T>
BST<T>::BST(T data)
{
    root = new_node(data);
}

template <class T>
BST<T>::BST(T * arr, int size)
{
    int i;

    root = new_node(arr[0]);

    for(i = 1; i < size; i++)
    {
        insert(arr[i]);
    }
}

template <class T>
BST<T>::~BST()
{
    destructor_a(root);
}

template <class T>
void BST<T>::destructor_a(Node * node)
{
    Node * right = NULL;

    if(node != NULL)
    {
        right = node->right;

        destructor_a(node->left);
        delete node;
        destructor_a(right);
    }
}

template <class T>
typename BST<T>::Node * BST<T>::get_min(Node * node)
{
    Node * min = NULL;

    while(node->left != NULL)
    {
        min = node->left;
    }

    return min;
}

template <class T>
void BST<T>::remove(T data)
{
    remove_a(data, root);
}

template <class T>
typename BST<T>::Node * BST<T>::remove_a(T data, Node * node)
{
    Node * temp = NULL;

    if(node == NULL)
    {
        return node;
    }

    if(data < node->data)
    {
        node->left = remove_a(data, node->left);
    }
    else if(data > node->data)
    {
        node->right = remove_a(data, node->right);
    }
    else
    {
        if(node->left == NULL)
        {
            temp = node->right;

            delete node;
            return temp;
        }
        else if(node->right == NULL)
        {
            temp = node->left;

            delete node;
            return temp;
        }

        temp = get_min(node->right);

        node->data = temp->data;
        node->right = remove_a(temp->data, node->right);
    }

    return node;
}

template <class T>
typename BST<T>::Node * BST<T>::new_node(T data)
{
    Node * node = new Node;

    node->data = data;
    node->left = NULL;
    node->right = NULL;

    return node;
}

template <class T>
void BST<T>::insert(T data)
{
    insert_a(data, root);
}

template <class T>
typename BST<T>::Node * BST<T>::insert_a(T data, Node * node)
{
    if(node == NULL)
    {
        return new_node(data);
    }

    if(data < node->data)
    {
        node->left = insert_a(data, node->left);
    }
    else if(data > node->data)
    {
        node->right = insert_a(data, node->right);
    }
}

template <class T>
void BST<T>::inorder()
{
    inorder_a(root);
}

template <class T>
void BST<T>::inorder_a(Node * node)
{
    if(node != NULL)
    {
        inorder_a(node->left);
        printf("%d \n", node->data);
        inorder_a(node->right);
    }
}

template <class T>
void BST<T>::postorder()
{
    postorder_a(root);
}

template <class T>
void BST<T>::postorder_a(Node * node)
{
    if(node != NULL)
    {
        postorder_a(node->right);
        printf("%d \n", node->data);
        postorder_a(node->left);
    }
}

#endif

File: dsa/python/BST.py
class BST:
    class Node:
        def __init__(self, data):
            self.data = data
            self.left = None
            self.right = None

    def __init__(self, data):
        self.root = self.Node(data)

    def find(self, key):
        current = self.root

        while current.data != key:
            if key < current.data:
                current = current.left
            else:
                current = current.right

            if current == None:
                return None

        return current

    def insert(self, key):
        node = self.Node(key)

        if self.root == None:
            self.root = node
            return

        current = self.root

        while True:
            parent = current

            if key == current.data:
                return
            elif key < current.data:
                current = current.left

                if current == None:
                    parent.left = node
                    return
            else:
                current = current.right

                if current == None:
                    parent.right = node
                    return

    def inorder(self):
        vals = []
        self.do_inorder(self.root, vals)
        return vals

    def do_inorder(self, node, vals):
        if node != None:
            self.do_inorder(node.left, vals)
            vals.append(node.data)
            self.do_inorder(node.right, vals)

    def postorder(self):
        vals = []
        self.do_postorder(self.root, vals)
        return vals

    def do_postorder(self, node, vals):
        if node != None:
            self.do_postorder(node.right, vals)
            vals.append(node.data)
            self.do_postorder(node.left, vals)

    def min(self):
        current = self.root
        last = None

        while current != None:
            last = current
            current = current.left

        return last

    def max(self):
        current = self.root
        last = None

        while current != None:
            last = current
            current = current.right

        return last

    def get_successor(self, node):
        parent = node
        successor = node
        current = node.right

        while current != None:
            parent = successor
            successor = current
            current = current.left

        if successor != node.right:
            parent.left = successor.right
            successor.right = node.right

        return successor

    def delete(self, key):
        current = self.root
        parent = self.root
        del_left = True

        while current.data != key:
            parent = current

            if key < current.data:
                del_left = True
                current = current.left
            else:
                del_left = False
                current = current.right

            if current == None:
                return False

        if current.left == None and current.right == None:
            if current == self.root:
                self.root = None
            elif del_left == True:
                parent.left = None
            else:
                parent.right = None
        elif current.right == None:
            if current == self.root:
                self.root = current.left
            elif del_left == True:
                parent.left = current.left
            else:
                parent.right = current.left
        elif current.left == None:
            if current == self.root:
                self.root = current.right
            elif del_left == True:
                parent.left = current.right
            else:
                parent.right = current.right
        else:
            successor = self.get_successor(current)

            if current == self.root:
                self.root = successor
            elif del_left == True:
                parent.left = successor
            else:
                parent.right = successor

            successor.left = current.left

        return True
